### PR TITLE
Alerting: Add actions extension point to alert instances table view

### PR DESCRIPTION
**Pull request by sd2k**
**What is this feature?**

This PR adds a new [extension point](https://grafana.com/developers/plugin-tools/ui-extensions/create-an-extension-point) to each row of the alert instances table, which is shown in both the 'Rule list' and the 'Rule details' pages. This allows plugins to add actions to a dropdown menu in the rightmost column of the table. Those actions are passed the alert instance so can use it for contextual handling. This is similar to the dropdown menu in the 'Explore' page (when you click 'Add') or in the panel menu dropdown (under 'Extensions').

See https://github.com/grafana/machine-learning/pull/3461 for an video example of how this can be used (e.g. by Grafana Sift here).

**Why do we need this feature?**

So far if plugins have wanted to add functionality to the alerting page it's been O(N) in the number of UI features added - e.g. adding the Grafana Incident button required work on the alerting page that was specific to that plugin. By adding a plugin extension point, any plugin can easily inject useful functionality in a non-intrusive way without needing further development on the alerting side.

**Who is this feature for?**

Plugin developers who would like to hook into the alerting page (specifically the Sift team, right now), and users who can then interact with alert instances in more varied ways.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/machine-learning/issues/3458.

**Special notes for your reviewer:**

Things to consider:

- UX: is it immediately obvious that the icon is a clickable button? what could we replace it with, if not?
  - I chose to omit a heading and keep the column nice and narrow so it was as unintrusive as possible, but it's not massively discoverable now 🤔 
- I've updated existing tests but haven't added any specifically for this, happy to do so but might need some guidance.
- The `PluginExtensionAlertInstanceContext` should probably be exported as part of `@grafana/data` (similar to `PluginExtensionPanelContext`), but it contains private types i.e. (`Alert`). Is there a subset of this `Alert` interface we could add to `@grafana/data`?

Screenshots:

unexpanded: 
image

expanded with one action:
image

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by sunker — COMMENTED** [review]
I think this looks great, but would like someone with more experience from working with extensions to have a look too. @mckn @leventebalogh can you review?

Regarding the icon button. This doesn't feel totally intuitive to me. Think there's an `ellipsis` icon in grafana ui. Can you try that instead?

**Comment by sd2k** [comment]
Sure! Here's how it looks with `ellipsis-v`:

image
image

**Comment by sunker** [comment]
> Sure! Here's how it looks with `ellipsis-v`:
> 
> image image

Think that looks much better, thank you!

**Comment by sympatheticmoose** [comment]
Screenshot 2023-11-10 at 13 06 49

In other extension points we're indicating which app has contributed the extension, maybe a pattern to follow here too?

**Comment by sd2k** [comment]
@sympatheticmoose we're doing that here too - that's what the 'Sift' part is in the second screenshot. It's a bit subtle because there's only one there!

(I actually imported the menu component directly from the explore toolbar extension point so it should be exactly the same)

**Comment by sympatheticmoose** [comment]
@sd2k d'oh, Friday eyes 😄

**Review by mckn — APPROVED** [review]
LGTM! We should probably wait for @grafana/alerting-squad prior to merging this one.

**Comment by gotjosh** [comment]
A tiny nit that perhaps the title should include "table view" - I got terrified for a second as I thought this was referring to the alert instance *database* table.

File: public/app/features/alerting/unified/components/extensions/AlertInstanceExtensionPoint.tsx
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/extensions/AlertInstanceExtensionPoint.tsx:30]: I believe we can improve this part of the UI; there's generally two patterns we tend to use in Alerting.

1. We don't show the dropdown at all if no actions can be taken, a disabled button with no explanation is a tad confusing :)
2. We always show the dropdown, with all of the actions but disable the actions the user has no permissions for are cannot take for some other reason.

If 2 we try to wrap the item with a tooltip explaining why we've disabled the action if we know the reason for it (insufficient permissions, x is not installed / configured, etc)

File: public/app/plugins/panel/alertlist/unified-alerting/GroupedView.tsx
Comment (gillesdemey) [inline public/app/plugins/panel/alertlist/unified-alerting/GroupedView.tsx:11]: nit: with the removal of this import we're no longer using the `GroupedRules` type anywhere from what I can tell. 

We should clean that up and remove the type :)

**Review by gillesdemey — COMMENTED** [review]
Code changes look good, thanks @sd2k! Couple of small nits and some changes on the UI / interaction model. 

Tagging @dhalachliyski here to drop some thoughts on that :)

**Review by leventebalogh — APPROVED** [review]
LGTM (mostly from a plugin-extensions perspective) 👍

File: public/app/features/alerting/unified/components/extensions/AlertInstanceExtensionPoint.tsx
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/extensions/AlertInstanceExtensionPoint.tsx:30]: If modifications need to happen to support the second pattern let's go with the first one to unblock this PR – that should be good from a UX perspective :)

**Comment by gillesdemey** [comment]
Code changes aside, does this warrant adding to the changelog / what's new if public plugins can inject actions? Do we need to update docs in this PR?

**Comment by sd2k** [comment]
> Code changes aside, does this warrant adding to the changelog / what's new if public plugins can inject actions? Do we need to update docs in this PR?

I guess that's an alerting decision, but from my perspective it might be nice to call out in the changelog that users may find some new actions in their alert instances lists, depending on the plugins they have installed. I'm not sure how that should be documented from a user's POV though, since we don't know what other plugins will extend it.

From a dev POV, maybe the plugin extension docs should be updated to reference the new extension point, but I'll leave that up to the plugin extensions folks to decide!

**Comment by sympatheticmoose** [comment]
> I'm not sure how that should be documented from a user's POV though, since we don't know what other plugins will extend it.

It _could_ be a fairly generic statement such as "Application plugins can contribute extensions which allow you to take actions based on a given alert, such as running a [Sift investigation](https://grafana.com/docs/grafana-cloud/alerting-and-irm/machine-learning/sift/) to help uncover the route cause of the alert" (I'm not sure on an appropriate location within the alerting docs) 

> From a dev POV, maybe the plugin extension docs should be updated to reference the new extension point, but I'll leave that up to the plugin extensions folks to decide!

Would welcome a PR to add it to the list here https://grafana.com/developers/plugin-tools/ui-extensions/register-an-extension#available-extension-points-within-grafana 🙏

**Comment by sd2k** [comment]
Thanks for all the reviews folks. @gillesdemey I'll leave the final decision to merge up to you - happy to contribute some docs if you'd like some, but I'm not 100% sure where they live!

@sympatheticmoose I've got a [draft PR](https://github.com/grafana/plugin-tools/pull/545) in place for the plugin tools docs 👍

**Review by gillesdemey — APPROVED** [review]
This is mostly for our own plugins for now – so I think it's fine to merge without docs update :)

LGTM! 🚀

**Pull request merged**
